File: typescript/rna-transcription/rna-transcription.ts
enum DNA {
    C = 'C',
    G = 'G',
    A = 'A',
    T = 'T',
}

const isDNA = (s: string) => Object.keys(DNA).includes(s)

enum RNA {
    G = 'G',
    C = 'C',
    U = 'U',
    A = 'A',
}

class Transcriptor {
    readonly transcript: {[key: string]: RNA} = {
        [DNA.C]: RNA.G,
        [DNA.G]: RNA.C,
        [DNA.A]: RNA.U,
        [DNA.T]: RNA.A
    }

    toRna(s: string): string {
        const nucleotides = s.split('')

        for (const strand of nucleotides) {
            if (!isDNA(strand)) {
                throw new Error('Invalid input DNA.')
            }
        }

        return nucleotides.map((strand) => this.transcript[strand]).join('')
    }
}

export default Transcriptor

File: kotlin/rna-transcription/src/main/kotlin/RnaTranscription.kt
enum class RNA(val value: Char) {
    G('G'),
    C('C'),
    A('A'),
    U('U');

    companion object {
        fun from(dna: DNA) =
            when (dna) {
                DNA.C -> G
                DNA.G -> C
                DNA.T -> A
                DNA.A -> U
            }
    }
}

enum class DNA(val value: Char) {
    C('C'),
    G('G'),
    T('T'),
    A('A');

    companion object {
        fun of(v: Char) = values().first { v == it.value }
    }
}

data class RnaStrands(val nucleotides: List<RNA>) {
    override fun toString() = nucleotides.map { it.value }.joinToString("")

    companion object {
        fun from(dnaStrands: DnaStrands) = dnaStrands.nucleotides.map { RNA.from(it) }.let(::RnaStrands)
    }
}

class DnaStrands(val nucleotides: List<DNA>) {
    companion object {
        fun of(v: String) = v.map { DNA.of(it) }.let(::DnaStrands)
    }
}

fun transcribeToRna(dna: String): String = DnaStrands.of(dna).let { RnaStrands.from(it) }.toString()
